pkg/reconciler/instances/istio/tests/integration_test.go

```diff
@@ -0,0 +1,167 @@
+package tests
+
+import (
+	"fmt"
+	"net/http"
+	"os"
+	"testing"
+
+	"github.com/avast/retry-go"
+	"github.com/stretchr/testify/require"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer/yaml"
+	"k8s.io/kubectl/pkg/util/podutils"
+)
+
+const (
+	istioNamespace         = "istio-system"
+	pilotIngressGwSelector = "istio in (pilot, ingressgateway)"
+	healthzTestURLFormat   = "http://%s:%s/healthz/ready"
+	gatewayManifest        = `
+apiVersion: networking.istio.io/v1beta1
+kind: Gateway
+metadata:
+  name: test-gateway
+spec:
+  selector:
+    istio: ingressgateway # use Istio default gateway implementation
+  servers:
+    - port:
+        number: 80
+        name: http
+        protocol: HTTP
+      hosts:
+        - "*"
+`
+	virutalServiceManifest = `
+apiVersion: networking.istio.io/v1beta1
+kind: VirtualService
+metadata:
+  name: istio-healthz
+spec:
+  gateways:
+  - istio-system/test-gateway
+  hosts:
+  - "*"
+  http:
+  - match:
+    - uri:
+        exact: /healthz/ready
+    route:
+    - destination:
+        host: istio-ingressgateway.istio-system.svc.cluster.local
+        port:
+          number: 15021
+`
+)
+
+var (
+	gatewayGVR = schema.GroupVersionResource{Resource: "gateways", Group: "networking.istio.io", Version: "v1beta1"}
+	vsGVR      = schema.GroupVersionResource{Resource: "virtualservices", Group: "networking.istio.io", Version: "v1beta1"}
+)
+
+func TestIstioIntegration(t *testing.T) {
+	skipTestIfDisabled(t)
+
+	setup := newIstioTest(t)
+	defer setup.contextCancel()
+
+	options := metav1.ListOptions{
+		LabelSelector: "app.kubernetes.io/instance=istio",
+	}
+
+	clientset, err := setup.kubeClient.Clientset()
+	require.NoError(t, err)
+
+	podsList, err := clientset.CoreV1().Pods(namespace).List(setup.context, options)
+	require.NoError(t, err)
+
+	for i, pod := range podsList.Items {
+		setup.logger.Infof("Pod %v is deployed", pod.Name)
+		require.Equal(t, v1.PodPhase("Running"), pod.Status.Phase)
+		ready := podutils.IsPodAvailable(&podsList.Items[i], 0, metav1.Now())
+		require.Equal(t, true, ready)
+	}
+
+	t.Run("istio pods should be running", func(t *testing.T) {
+		pilotSelectorOpt := metav1.ListOptions{LabelSelector: pilotIngressGwSelector}
+		deploymentList, err := clientset.AppsV1().Deployments(istioNamespace).List(setup.context, pilotSelectorOpt)
+		require.NoError(t, err)
+		require.True(t, len(deploymentList.Items) >= 2)
+		for _, dep := range deploymentList.Items {
+			require.Equal(t, *dep.Spec.Replicas, dep.Status.ReadyReplicas)
+			require.Equal(t, *dep.Spec.Replicas, dep.Status.AvailableReplicas)
+		}
+	})
+
+	t.Run("healthz returns 200", func(t *testing.T) {
+		gateway := &unstructured.Unstructured{}
```

Comment:
Wouldn't it be a better idea to return a newly-created unstructured object from the function instead of passing a reference to it? With a return value the contract is clear at first glance. Now it works as a side-effect when you deserialise a yaml into an object